runtime/init-container/src/init.c

```diff
@@ -1993,25 +2045,22 @@ int main(int argc, char **argv) {
         fprintf(stderr, "Environment variable: %s\n", *p);
     }
 
-    glob_t nvidia_nodes;
-    res = glob("/dev/nvidia[0-9]*", GLOB_ERR, NULL, &nvidia_nodes);
-    if (res == 0) {
+    if (nvidia_loaded) {
         if (do_sandbox == false) {
             fprintf(stderr, "Sandboxing is disabled, refusing to enable Nvidia GPU passthrough.\n");
             fprintf(stderr, "Please re-run the container with sandboxing enabled or disable GPU passthrough.\n");
             errno = 0;
             CHECK_BOOL(0);
         }
-        struct stat statbuf;
-        for (size_t i = 0; i < nvidia_nodes.gl_pathc; ++i) {
-            CHECK_BOOL(strncmp(nvidia_nodes.gl_pathv[i], "/dev/nvidia", sizeof "/dev/nvidia" - 1) == 0);
-            CHECK_BOOL(stat(nvidia_nodes.gl_pathv[i], &statbuf) == 0);
-            CHECK_BOOL(S_ISCHR(statbuf.st_mode));
-            res = mknodat(g_sysroot_fd, nvidia_nodes.gl_pathv[i] + 1, statbuf.st_dev, statbuf.st_mode & 0777);
-            CHECK_BOOL(res == 0 || (res == -1 && errno == EEXIST));
-        }
-    } else {
-        CHECK_BOOL(res == GLOB_NOMATCH);
+        int nvidia_major = CHECK(find_device_major("nvidia-frontend"));
+        /* TODO: multi-card support needs more /dev/nvidia%d nodes */
+        res = mknodat(g_sysroot_fd, "/dev/nvidia0", S_IFCHR | (0666 & 0777), nvidia_major << 8 | 0);
+        CHECK_BOOL(res == 0 || (res == -1 && errno == EEXIST));
+        res = mknodat(g_sysroot_fd, "/dev/nvidiactl", S_IFCHR | (0666 & 0777), nvidia_major << 8 | 255);
+        CHECK_BOOL(res == 0 || (res == -1 && errno == EEXIST));
+        nvidia_major = CHECK(find_device_major("nvidia-uvm"));
+        res = mknodat(g_sysroot_fd, "/dev/nvidia-uvm", S_IFCHR | (0666 & 0777), nvidia_major << 8 | 0);
+        CHECK_BOOL(res == 0 || (res == -1 && errno == EEXIST));
```

Comment:
Leading `/` means that the file descriptor will be ignored and the paths created outside the chroot, which is wrong.